File: src/Integra/Token.hs
module Integra.Token (Token(..), lexer) where

import Data.Char (isDigit, isSpace, isAlpha)

data Token
    = NumberTok Double
    | PlusTok | MinusTok | TimesTok | DivideTok | PowerTok
    | LParenTok | RParenTok
    | VarTok
    | SinTok | CosTok | TanTok
    | CscTok | SecTok | CotTok
    | AsinTok | AcosTok | AtanTok
    | AcscTok | AsecTok | AcotTok
    | SinhTok | CoshTok | TanhTok
    | CschTok | SechTok | CothTok
    | AsinhTok | AcoshTok | AtanhTok
    | LogTok | Log2Tok | Log10Tok | Log1pTok
    | ExpTok | Expm1Tok | SqrtTok | CbrtTok
    | AbsTok | SignTok
    | FloorTok | CeilTok | RoundTok
    | GammaTok | ErfTok
    | ConjTok | ReTok | ImTok
    | PiTok | ETok | TauTok | PhiTok | ITok
    | AnsTok
    | GreaterTok | GreaterEqTok | LessTok | LessEqTok | EqTok | NeqTok
    deriving (Show, Eq)

lexer :: String -> [Token]
lexer [] = []
lexer (x:xs) | isSpace x = lexer xs
lexer ('+':xs) = PlusTok       : lexer xs
lexer ('-':xs) = MinusTok      : lexer xs
lexer ('*':xs) = TimesTok      : lexer xs
lexer ('/':xs) = DivideTok     : lexer xs
lexer ('^':xs) = PowerTok      : lexer xs
lexer ('(':xs) = LParenTok     : lexer xs
lexer (')':xs) = RParenTok     : lexer xs
lexer ('>':'=':xs) = GreaterEqTok : lexer xs
lexer ('>':xs)     = GreaterTok    : lexer xs
lexer ('<':'=':xs) = LessEqTok    : lexer xs
lexer ('<':xs)     = LessTok       : lexer xs
lexer ('=':'=':xs) = EqTok         : lexer xs
lexer ('!':'=':xs) = NeqTok        : lexer xs
lexer ('l':'o':'g':'1':'0':xs) = Log10Tok : lexer xs
lexer ('l':'o':'g':'1':'p':xs) = Log1pTok : lexer xs
lexer ('l':'o':'g':'2':xs)     = Log2Tok  : lexer xs
lexer ('e':'x':'p':'m':'1':xs) = Expm1Tok : lexer xs
lexer (x:xs) | isDigit x || x == '.' =
    let (num, rest) = span (\c -> isDigit c || c == '.') (x:xs)
    in NumberTok (read num) : lexer rest
lexer (x:xs) | isAlpha x =
    let (name, rest) = span isAlpha (x:xs)
    in case name of
        "sin"   -> SinTok   : lexer rest
        "cos"   -> CosTok   : lexer rest
        "tan"   -> TanTok   : lexer rest
        "csc"   -> CscTok   : lexer rest
        "sec"   -> SecTok   : lexer rest
        "cot"   -> CotTok   : lexer rest
        "asin"  -> AsinTok  : lexer rest
        "acos"  -> AcosTok  : lexer rest
        "atan"  -> AtanTok  : lexer rest
        "acsc"  -> AcscTok  : lexer rest
        "asec"  -> AsecTok  : lexer rest
        "acot"  -> AcotTok  : lexer rest
        "sinh"  -> SinhTok  : lexer rest
        "cosh"  -> CoshTok  : lexer rest
        "tanh"  -> TanhTok  : lexer rest
        "csch"  -> CschTok  : lexer rest
        "sech"  -> SechTok  : lexer rest
        "coth"  -> CothTok  : lexer rest
        "asinh" -> AsinhTok : lexer rest
        "acosh" -> AcoshTok : lexer rest
        "atanh" -> AtanhTok : lexer rest
        "ln"    -> LogTok   : lexer rest
        "log"   -> LogTok   : lexer rest
        "log1p" -> Log1pTok : lexer rest
        "exp"   -> ExpTok   : lexer rest
        "expm1" -> Expm1Tok : lexer rest
        "sqrt"  -> SqrtTok  : lexer rest
        "cbrt"  -> CbrtTok  : lexer rest
        "abs"   -> AbsTok   : lexer rest
        "sign"  -> SignTok  : lexer rest
        "floor" -> FloorTok : lexer rest
        "ceil"  -> CeilTok  : lexer rest
        "round" -> RoundTok : lexer rest
        "gamma" -> GammaTok : lexer rest
        "erf"   -> ErfTok   : lexer rest
        "conj"  -> ConjTok  : lexer rest
        "re"    -> ReTok    : lexer rest
        "im"    -> ImTok    : lexer rest
        "pi"    -> PiTok    : lexer rest
        "tau"   -> TauTok   : lexer rest
        "phi"   -> PhiTok   : lexer rest
        "e"     -> ETok     : lexer rest
        "inf"   -> NumberTok (1/0) : lexer rest
        "i"     -> ITok     : lexer rest
        "x"     -> VarTok   : lexer rest
        "ans"   -> AnsTok   : lexer rest
        _       -> error ("Unknown identifier: " ++ name)
lexer (x:_) = error ("Unknown character: " ++ [x])

File: wasm/src/token.rs
#[derive(Clone, Debug, PartialEq)]
pub enum Token {
    Num(f64),
    Plus,
    Minus,
    Times,
    Divide,
    Power,
    LParen,
    RParen,
    Var,
    Ans,
    Sin,
    Cos,
    Tan,
    Csc,
    Sec,
    Cot,
    Asin,
    Acos,
    Atan,
    Acsc,
    Asec,
    Acot,
    Sinh,
    Cosh,
    Tanh,
    Csch,
    Sech,
    Coth,
    Asinh,
    Acosh,
    Atanh,
    Log,
    Log2,
    Log10,
    Log1p,
    Exp,
    Expm1,
    Sqrt,
    Cbrt,
    Abs,
    Sign,
    Floor,
    Ceil,
    Round,
    Gamma,
    Erf,
    Conj,
    Re,
    Im,
    Pi,
    E,
    Tau,
    Phi,
    I,
    Greater,
    GreaterEq,
    Less,
    LessEq,
    Eq,
    Neq,
}

pub fn tokenize(s: &str) -> Vec<Token> {
    let s = s.as_bytes();
    let mut t: Vec<Token> = Vec::new();
    let mut i = 0;
    while i < s.len() {
        if s[i] == b' ' || s[i] == b'\t' {
            i += 1;
            continue;
        }
        if s[i] == b'+' { t.push(Token::Plus); i += 1; continue; }
        if s[i] == b'-' { t.push(Token::Minus); i += 1; continue; }
        if s[i] == b'*' { t.push(Token::Times); i += 1; continue; }
        if s[i] == b'/' { t.push(Token::Divide); i += 1; continue; }
        if s[i] == b'^' { t.push(Token::Power); i += 1; continue; }
        if s[i] == b'(' { t.push(Token::LParen); i += 1; continue; }
        if s[i] == b')' { t.push(Token::RParen); i += 1; continue; }

        if s[i] == b'>' {
            if i + 1 < s.len() && s[i + 1] == b'=' {
                t.push(Token::GreaterEq); i += 2; continue;
            }
            t.push(Token::Greater); i += 1; continue;
        }
        if s[i] == b'<' {
            if i + 1 < s.len() && s[i + 1] == b'=' {
                t.push(Token::LessEq); i += 2; continue;
            }
            t.push(Token::Less); i += 1; continue;
        }
        if s[i] == b'=' && i + 1 < s.len() && s[i + 1] == b'=' {
            t.push(Token::Eq); i += 2; continue;
        }
        if s[i] == b'!' && i + 1 < s.len() && s[i + 1] == b'=' {
            t.push(Token::Neq); i += 2; continue;
        }

        if s[i] == b'.' || (s[i] >= b'0' && s[i] <= b'9') {
            let start = i;
            while i < s.len() && (s[i] >= b'0' && s[i] <= b'9' || s[i] == b'.') {
                i += 1;
            }
            let num_str = std::str::from_utf8(&s[start..i]).unwrap();
            let v: f64 = num_str.parse().unwrap_or(0.0);
            t.push(Token::Num(v));
            continue;
        }

        if (s[i] >= b'a' && s[i] <= b'z') || (s[i] >= b'A' && s[i] <= b'Z') {
            let start = i;
            while i < s.len()
                && ((s[i] >= b'a' && s[i] <= b'z')
                    || (s[i] >= b'A' && s[i] <= b'Z')
                    || (s[i] >= b'0' && s[i] <= b'9'))
            {
                i += 1;
            }
            let id = std::str::from_utf8(&s[start..i]).unwrap();
            match id {
                "sin" => t.push(Token::Sin),
                "cos" => t.push(Token::Cos),
                "tan" => t.push(Token::Tan),
                "csc" => t.push(Token::Csc),
                "sec" => t.push(Token::Sec),
                "cot" => t.push(Token::Cot),
                "asin" => t.push(Token::Asin),
                "acos" => t.push(Token::Acos),
                "atan" => t.push(Token::Atan),
                "acsc" => t.push(Token::Acsc),
                "asec" => t.push(Token::Asec),
                "acot" => t.push(Token::Acot),
                "sinh" => t.push(Token::Sinh),
                "cosh" => t.push(Token::Cosh),
                "tanh" => t.push(Token::Tanh),
                "csch" => t.push(Token::Csch),
                "sech" => t.push(Token::Sech),
                "coth" => t.push(Token::Coth),
                "asinh" => t.push(Token::Asinh),
                "acosh" => t.push(Token::Acosh),
                "atanh" => t.push(Token::Atanh),
                "ln" => t.push(Token::Log),
                "log1p" => t.push(Token::Log1p),
                "log10" => t.push(Token::Log10),
                "log2" => t.push(Token::Log2),
                "log" => t.push(Token::Log),
                "exp" => t.push(Token::Exp),
                "expm1" => t.push(Token::Expm1),
                "sqrt" => t.push(Token::Sqrt),
                "cbrt" => t.push(Token::Cbrt),
                "abs" => t.push(Token::Abs),
                "sign" => t.push(Token::Sign),
                "floor" => t.push(Token::Floor),
                "ceil" => t.push(Token::Ceil),
                "round" => t.push(Token::Round),
                "gamma" => t.push(Token::Gamma),
                "erf" => t.push(Token::Erf),
                "conj" => t.push(Token::Conj),
                "re" => t.push(Token::Re),
                "im" => t.push(Token::Im),
                "pi" => t.push(Token::Pi),
                "tau" => t.push(Token::Tau),
                "phi" => t.push(Token::Phi),
                "inf" => t.push(Token::Num(f64::INFINITY)),
                "e" => t.push(Token::E),
                "i" => t.push(Token::I),
                "x" => t.push(Token::Var),
                "ans" => t.push(Token::Ans),
                _ => {} // ignore unknown identifiers
            }
            continue;
        }

        i += 1;
    }
    t
}
